src/class/vendor/vendor_device.c

```diff
@@ -247,38 +287,33 @@ uint16_t vendord_open(uint8_t rhport, const tusb_desc_interface_t* desc_itf, uin
 
 bool vendord_xfer_cb(uint8_t rhport, uint8_t ep_addr, xfer_result_t result, uint32_t xferred_bytes) {
   (void) result;
+  const uint8_t idx = find_vendor_itf(ep_addr);
+  TU_VERIFY(idx < CFG_TUD_VENDOR);
+  vendord_interface_t *p_vendor = &_vendord_itf[idx];
 
-  uint8_t itf;
-  vendord_interface_t* p_vendor;
-
-  for (itf = 0; itf < CFG_TUD_VENDOR; itf++) {
-    p_vendor = &_vendord_itf[itf];
-    if ((ep_addr == p_vendor->rx.stream.ep_addr) || (ep_addr == p_vendor->tx.stream.ep_addr)) {
-      break;
-    }
-  }
-  TU_VERIFY(itf < CFG_TUD_VENDOR);
-  vendord_epbuf_t* p_epbuf = &_vendord_epbuf[itf];
-
-  if ( ep_addr == p_vendor->rx.stream.ep_addr ) {
+  if (ep_addr == p_vendor->stream.rx.ep_addr) {
     // Received new data: put into stream's fifo
-    tu_edpt_stream_read_xfer_complete(&p_vendor->rx.stream, xferred_bytes);
+    tu_edpt_stream_read_xfer_complete(&p_vendor->stream.rx, xferred_bytes);
 
-    // Invoked callback if any
-    tud_vendor_rx_cb(itf, p_epbuf->epout, (uint16_t) xferred_bytes);
+    // invoke callback
+  #if CFG_TUD_VENDOR_RX_BUFSIZE == 0
+    tud_vendor_rx_cb(idx, p_vendor->stream.rx.ep_buf, xferred_bytes);
+  #else
+    tud_vendor_rx_cb(idx, NULL, 0);
+  #endif
 
-    tu_edpt_stream_read_xfer(rhport, &p_vendor->rx.stream);
-  } else if ( ep_addr == p_vendor->tx.stream.ep_addr ) {
+    tu_edpt_stream_read_xfer(rhport, &p_vendor->stream.rx); // prepare next data
```

Comment:
I think it's useful to add a config whether next transfer is scheduled automatically or manually later in `read_finished()`.

To give processing time for some commands.